Update release process for Providers (#24680)
mistercrunch/Airflow: README.md

to 2.3.0 in the first Provider's release after 11th of October 2022 (11th of October 2021 is the date when the
first `PATCHLEVEL` of 2.2 (2.2.0) has been released.

Since providers are often connected with some stakeholders that are vitally interested in maintaining
their integrations (for example cloud providers, ore specific service providers), and we are also bound with
the Apache Software Foundation release policies. The provider's governance model is something we name

> "mixed governance".

Reviewer: - Please clarify "we are also bound with the Apache Software Foundation release policies"
- "providers are often connected with some stakeholders" are those stakeholders documented somewhere?
- Please clarify "The provider's governance model"
- You throw in the term "mixed governance" but don't refer to it anywhere. Can you elaborate on it?

Author: > * Please clarify "we are also bound with the Apache Software Foundation release policies"

Added link and describe that it's about "who and how releases the ASF software"

> * "providers are often connected with some stakeholders" are those stakeholders documented somewhere?

No, but you do not have to be a stakeholder to propose to release an older version of a provider if you wish to take the burden of cherry-picking and testing the changes. So I do not think we need to document those. Just a PR to an old branch and confirmation that it's been tested is enough.

> * Please clarify "The provider's governance model"

Did.

> * You throw in the term "mixed governance" but don't refer to it anywhere. Can you elaborate on it?

Did.

The Airflow Community and release manager still decides when to release those providers.

> This is fully managed by the community and the usual release-management process.

Reviewer: IMO this paragraph doesn't hold any useful information. As a reader, in case I would have a question; what are useful resources or channels to ask?

Author: This is (not very precisely explained but I just changed it) the "governance model". And I've added the link to the ASF relese policies that not only explain the mechanics, reasoning but also  have a bunch of FAQs to answer.

Usually, we release only the most recent version of the provider and rather aggressively
remove deprecations in "major" versions of the providers, however the stakeholder in a given provider might
agree and make their effort on cherry-picking the non-breaking changes to a selected previous major branch
of the provider which results in releasing more (usually two) versions of such provider when we release it:
potentially breaking "latest" major version, and selected past major version with non-breaking changes

> applied by the stakeholder (cherry-picked changes have to be merged by the committer following the usual

Reviewer 2: > cherry-picked changes have to be merged by the committer following the usual rules of the community

And by this you mean proposing the changes (e.g. cherry-picked changes from main) in a PR from a fork, right?

Author: Correct. As described in the comment to  @BasPH comments -  I want to describe technicalities when we do it for th first time in a "dev" process.

> rules of the community).

Reviewer: - "however the stakeholder in a given provider might agree" agree with what?
- Can you give an example of when this happens, e.g. critical bug or security fix?

Author: > * "however the stakeholder in a given provider might agree" agree with what?

Clarified that this does not have to be a stakeholder and remove the "agree" - it was meaningless.

> * Can you give an example of when this happens, e.g. critical bug or security fix?

Yeah.  I tried to describe it better. It's really about "coupling" - the current "release only latest" provider effectively couples important bug-fixes with some unrelated breaking changes. The approach I propose allows to decouple those.

Author: I could add more specific examples from the past, but I think it's a bit too much for such "policy".

The community continues to release such older versions of the providers for as long as there is an effort
of the stakeholder to maintain the cherry-picked changes. The availability of the stakeholder that can
manage "service-oriented" maintenance and agrees to such a responsibility, will also drive our willingness

> to accept future, new providers to become community managed.

Reviewer: Not sure what message you want to convey here. Is this paragraph necessary?

Reviewer 2: I think this is a very important paragraph, and if anything, should be expanded! And also made more concrete.

> The community continues to release such older versions of the providers for as long as there is an effort
of the stakeholder to maintain the cherry-picked changes.

This leaves the door open to having some _very_ old actrive branches :grin: 

> The availability of the stakeholder that can manage "service-oriented" maintenance and agrees to such a responsibility, will also drive our willingness to accept future, new providers to become community **managed.**

We should formalize what level of availability is required and make it concrete what such an agreement looks like. Is that a PR to a piece of code/documentation? A chain on the dev list? Something else entirely?

Author: > Not sure what message you want to convey here. Is this paragraph necessary?

This is all about managing the expectations of the users. What we are writing here as a policy is a policy that our users will be able (and they will do it) to call us upon. This happened multiple times ("but you wrote here that you are doing it"). The message was that the users should not *expect* older versions to be released - if there is no-one to cherry-pick them we (as maintainers) will not be expected to release the older versions. I updated it slighttly

> I think this is a very important paragraph, and if anything, should be expanded! And also made more concrete.

I did.

> This leaves the door open to having some very old actrive branches grin

True. But I added "to perform the cherry-picks and carry-on testing of the older provider version." - the testing part will limit it significantly. I also changed in the previous paragrapsh "usually two version" to "maximum two versions". That should help.

> We should formalize what level of availability is required and make it concrete what such an agreement looks like. Is that a PR to a piece of code/documentation? A chain on the dev list? Something else entirely?

I think there wil never be formal agreement - this will always be slightly "vague" statement. Driving willingess is gooo enough IMHO as an indication that there is a relation between the availabilihty and likelihood of being accepted. From the community side - everything in the ASF and "Apache Way" it's all driven by consensus - if we cannot achieve, we vote. From the "stakeholder" point of view - this is "well, when I contribute and want to release older versions, I or others will have to commit to cherry-pick to and test the older versions. This is the "responsibility" I am committing to.

to 2.3.0 in the first Provider's release after 11th of October 2022 (11th of October 2021 is the date when the
first `PATCHLEVEL` of 2.2 (2.2.0) has been released.

Providers are often connected with some stakeholders that are vitally interested in maintaining backwards
compatibilities in their integrations (for example cloud providers, or specific service providers). But,
we are also bound with the [Apache Software Foundation release policy](https://www.apache.org/legal/release-policy.html)

> which describes who and how releases the ASF software. The provider's governance model is something we name

Reviewer: ```suggestion
which describes who releases, and how to release the ASF software. The provider's governance model is something we name
```

"mixed governance" - where we follow the release policies, while the burden of maintaining and testing
the cherry-picked versions is on those who commit to perform the cherry-picks and make PRs to older
branches.

The "mixed governance" means that:

> * The Airflow Community and release manager still decides when to release those providers.

Reviewer: ```suggestion
* The Airflow Community and release manager decide when to release those providers.
```

  This is fully managed by the community and the usual release-management process following the
  [Apache Software Foundation release policy](https://www.apache.org/legal/release-policy.html)
* The contributors (who might or might not be direct stakeholders in the provider) will carry the burden
  of cherry-picking and testing the older versions of providers.

Usually, we release only the most recent version of the provider and rather aggressively
remove deprecations in "major" versions of the providers, however if there is a contributor willing to
make their effort on cherry-picking and testing the non-breaking changes to a selected previous major branch

> of the provider which results in releasing maximum two versions of such provider when we release it:

Reviewer: This whole paragraph is a single sentence and I don't understand it. Could you split into into multiple sentences for readability? For example:

> Usually, we release only the most recent version of the provider and rather aggressively
remove deprecations in "major" versions of the providers. However, ....

Author: Will do. It's called train of thoughts :)

Author: I restructured this paragraph.

BTW. Communicating your thoughts in clear way is HARD. I've literally read today about the Feynmann technique of learning https://fs.blog/feynman-technique/  and I think I am very far from my science guru :D

* potentially breaking "latest" major version
* selected past major version with non-breaking changes applied by the contributor

> Cherry-picking such changes follows the process that we already follow for releasing Airflow

Reviewer: ```suggestion
Cherry-picking such changes follows the same process for releasing Airflow
```

> patch-level releases for previous minor Airflow version. Usually such cherry-picking is done when
> there is an important bugfix but the latest version contains breaking changes which are not
> coupled with the bugfix (but releasing them together in the latest version of provider effectively couples
> them). The cherry-picked changes have to be merged by the committer following the usual rules of the

Reviewer: ```suggestion
patch-level releases for a previous minor Airflow version. Usually such cherry-picking is done when
there is an important bugfix and the latest version contains breaking changes that are not
coupled with the bugfix. Releasing them together in the latest version of the provider effectively couples
them, and therefore they're released separately. The cherry-picked changes have to be merged by the committer following the usual rules of the
```

to 2.3.0 in the first Provider's release after 11th of October 2022 (11th of October 2021 is the date when the
first `PATCHLEVEL` of 2.2 (2.2.0) has been released.

Providers are often connected with some stakeholders that are vitally interested in maintaining backwards
compatibilities in their integrations (for example cloud providers, or specific service providers). But,
we are also bound with the [Apache Software Foundation release policy](https://www.apache.org/legal/release-policy.html)
which describes who releases, and how to release the ASF software. The provider's governance model is something we name
"mixed governance" - where we follow the release policies, while the burden of maintaining and testing

> the cherry-picked versions is on those who commit to perform the cherry-picks and make PRs to older

Reviewer 3: minor suggestion - "while the responsibility of deciding cherry-picks for the older branches of a provider is on those who take ownership to maintain and test the cherry-picks against older versions of the given provider."

Author: Well. It's not the responsibility to decide. Maintainer decides if the PR is merged (as usual - no chang here). This is the "burden" - cherry-picking is not easy. This is mostly dirty, boring and ground work to be done, nothing really "cool". I've been doing it for years, so I know what I am talking about :). 

I want to be absolutely sure that this is well communicated. It's not a priviledge or ownership. This is committing to the hard work, without taking the ownership. The ownership and decision making stays with maintainers, but they want to spend as little time an energy as possible on doing it. The maintainer migth look at the PR and tell "this should not be cherry-picked" and it will not be merged. (this is how ASF rules work0. 

So summariising "responsibility to decide" and "ownership" are bad words in this context.

BTW. One of the purposes here is to make it very clear, that it's not a burden of the maintainers to make and you have to REALLY want it to commit and to do it. And by doing it, you materialise your commitment to take the burden of it.  There should be very few of those and only when there are really good reasons and a lot of effort put by those want to make it happen (just wanting is not envough).

Reviewer 3: > This is mostly dirty, boring and ground work to be done, nothing really "cool". I've been doing it for years, so I know what I am talking about :)

Fair enough :) Agree with your current language given the context you shared.

> branches.

Reviewer 3: Should we describe here that "mixed governance" model is opt-in for a provider? I am guessing this model will be followed only if there are stakeholders (or contributors) who are taking the responsibility to perform the cherry-picks and test them thoroughly.

Author: Yeah. Good point.

Author: I added this.

The "mixed governance" means that:

* The Airflow Community and release manager decide when to release those providers.
  This is fully managed by the community and the usual release-management process following the
  [Apache Software Foundation release policy](https://www.apache.org/legal/release-policy.html)
* The contributors (who might or might not be direct stakeholders in the provider) will carry the burden

>   of cherry-picking and testing the older versions of providers.

Reviewer 3: I am curious about a few things here - 
1. What would be the process for selecting these contributors? Would we ask the contributors to share detailed result of their testing (system tests) to confirm their claims?
2. Would this selection of contributors happen before every release? I am guessing, yes.

Author: 1. Just the usual in OSS - if someone wants it, they raise a PR or open issue. I want to describe it in details when we start doing it (we need to verify the process) - but I think. as usual anyone when their (or other's) PR gets merged might raise their hand and comment "hey - I want to cherry-pick this and other changes and test it for this and that version and I would like to get a branch I could reise resulting PR against. As usual in oss - if you want something to happen - take a lead and start doing it. That will also naturally filter the people who would "like" to have it but they would expect it will "magicallly happen". If you do, it , raise a good PR and cherry-pick all changes and confirm you tested the result of your PR and this is the 2nd provider version (not 3rd or 4th), this is fine. And anyone other can cherry-pick more PRS as well. No limits there - but someone must volunteer and commit to making it ready for releas to make it happen.

2. No selection. Actions by those people will decide (unless there will be conflicting requests which we can handle when they happen). for now I optimistically assume that there will be a lot of people who would "want" tbis to happen, there will be very few who will "make it happen" and there will be no conflicts. And we handle similar conflidcts on a daily basis when for example 2 people contribute the same bugfix or feature. It's far easier to manage it when it happens than to make a process to make sure it does not happen.

Reviewer 3: Thanks for clarifying. 
1. Aligning this with the current PR process makes complete sense. Can't wait to see this process playing out. 

2. The answer to the 1st question that the cherry-picking will go hand in hand with the PR process makes it very clear. You're right, we can resolve the conflicts as they arise, instead of documenting solution path for every scenario. 
> there will be a lot of people who would "want" this to happen, there will be very few who will "make it happen" 

I very much concur on this.

Author: I added explanation on that. Very good points @shubham22 !

Reviewer 3: The changes look great and will provide the reader and contributors clear insight into the process. Thank you!
